### PR TITLE
Complete CSE2000 contributions

### LINKED ISSUE #12: Add support for block-diagonal Hessian approximations
*This is an issue reserved for the [TU Delft Student Software Project '23 - LaplaceRedux](https://github.com/orgs/JuliaTrustworthyAI/projects/5/)*

Again a larger piece of work ...

Take a look at the Python [package](https://aleximmer.github.io/Laplace/) and note that they have implemented classes for different ways to approximate the Hessian (e.g. Laplace approximation with Kronecker factored log likelihood Hessian approximation). Our package currently only supports Empirical Fisher and you may have already tackled #17 to implement GGN. Both of these approaches approximate the full Hessian, which is costly. 

The goal here is to implement more approaches.

### PR DESCRIPTION
**Pull request by severinbratus**
Resolves #12, #17, #18, #19, #20, #27

Note, for #12, only KFAC for multi-class is included. KFAC is not applicable to regression. KFLR would require access to pre-activation values, which Zygote does not provide.

This PR includes an extensive and systematic test-suite covering most combinations of backends (`EmpiricalFisher` or `GGN`), network subset types (`all`, `subnetwork`, `lastlayer`), batchsizes, and Hessian structures (`full` or `kron`), comparing Hessians computed to those of `laplace-torch`.

We have also included some additions to the pipeline, namely: a formatting check, and a benchmark pipeline.

### CODE REVIEW

**Review by pat-alt — COMMENTED** [review]
Thanks all for your great work 🥳 

I'll merge this once I've read your report and attended your presentation. Will then also publish a new package release.

**Review by pat-alt — COMMENTED** [review]
Hi team, 

As far as I can tell from this, you haven't added any updates to the actual `/docs/`, is that right? In that case, what constitutes your documentation? Just everything that's under `dev`?

**Comment by pat-alt** [comment]
Some other things to address before we merge (TBC):

- [ ] missing docstrings in a few places

File: src/curvature/functions.jl [hunk 1]
```diff
@@ -26,89 +69,463 @@ end
...
 
-# "Constructor for Generalized Gauss Newton."
-# struct GGN <: CurvatureInterface
-#     model::Any
-#     likelihood::Symbol
-#     loss_fun::Function
-#     params::AbstractArray
-#     factor::Union{Nothing,Real}
-# end
+"Constructor for Empirical Fisher."
```
Comment (pat-alt) [inline src/curvature/functions.jl:80]: Should this no be GGN?

File: src/curvature/functions.jl [hunk 2]
```diff
@@ -26,89 +69,463 @@ end
...
+        full_unbatched(curvature, d)
+    end
+end
+
+"""
+    full(curvature::EmpiricalFisher, d::Union{Tuple,NamedTuple})
+
+Compute the full empirical Fisher for batch of inputs-outputs, with the batch dimension at the end.
+"""
+function full_unbatched(curvature::EmpiricalFisher, d::Tuple)
```
Comment (pat-alt) [inline src/curvature/functions.jl:216]: Wrong docstring?

File: src/curvature/functions.jl [hunk 3]
```diff
@@ -26,89 +69,463 @@ end
...
 
     # Empirical Fisher:
+    # - the product of the gradient vector with itself transposed
     H = 𝐠 * 𝐠'
-    
+
+    return loss, H
+end
+
+function full_batched(curvature::EmpiricalFisher, d::Tuple)
```
Comment (pat-alt) [inline src/curvature/functions.jl:236]: missing docstring

File: src/curvature/functions.jl [hunk 4]
```diff
@@ -26,89 +69,463 @@ end
...
+    end
+
+    # Empirical Fisher:
+    # H = 𝐠 * 𝐠'
+    @tullio H[i, j] := 𝐠[i, b] * 𝐠[j, b]
+
     return loss, H
+end
+
+"""
```
Comment (pat-alt) [inline src/curvature/functions.jl:257]: very cheeky 😆

File: src/curvature/functions.jl [hunk 5]
```diff
@@ -26,89 +69,463 @@ end
...
+macro zb(expr)
+    if expr.head == :ref
+        a, i = expr.args
+        return Expr(:ref, esc(a), Expr(:call, :+, esc(i), 1))
+    else
+        error("Expected expr of form a[i], got: ($(expr)) with head $(expr.head)")
+    end
+end
+
+struct Kron
```
Comment (pat-alt) [inline src/curvature/functions.jl:269]: Maybe I'm missing something (or can't find the corresponding docs), but the whole KFAC is missing docstrings here, that clarify what's happening.

File: src/curvature/functions.jl [hunk 6]
```diff
@@ -26,89 +69,463 @@ end
...
+"""
+Interleave elements of multiple iterables in order provided.
+"""
+function interleave(iters...)
+    return (elem for pack in zip(iters...) for elem in pack)
+end
+
+"""
+Compute KFAC for the Fisher.
+"""
```
Comment (pat-alt) [inline src/curvature/functions.jl:305]: The comments here are very helpful, but worth summarising in docstring

**Comment by severinbratus** [comment]
> Hi team,
> 
> As far as I can tell from this, you haven't added any updates to the actual `/docs/`, is that right? In that case, what constitutes your documentation? Just everything that's under `dev`?

Hi! This was addressed on the final presentation, but for the record: no, we did not update `docs/`, the supporting documentation we provide is in docstrings and notebooks. Sorry for the oversight with docstrings for KFAC, we'll fix that soon enough.

**Pull request merged**